lib/internal/modules/esm/assert.js

```diff
@@ -0,0 +1,57 @@
+'use strict';
+
+const {
+  SafeMap,
+  Symbol,
+  ObjectPrototypeHasOwnProperty,
+} = primordials;
+
+const {
+  ERR_IMPORT_ASSERTION_TYPE_FAILED,
+  ERR_IMPORT_ASSERTION_TYPE_UNSUPPORTED,
+} = require('internal/errors').codes;
+
+const kImplicitAssertType = Symbol('implicit assert type');
+
+/**
+ * Define a map of module formats to import assertion types (the value of `type`
+ * in `assert { type: 'json' }`).
+ * @type {Map<string, string | kImplicitAssertType}
+ */
+const formatTypeMap = new SafeMap([
+  ['builtin', kImplicitAssertType],
+  ['commonjs', kImplicitAssertType],
+  ['module', kImplicitAssertType],
+  ['json', 'json'],
+  ['wasm', kImplicitAssertType], // Should probably be 'webassembly' per https://github.com/tc39/proposal-import-assertions
+]);
+
+
+function validateAssertions(format, importAssertions) {
+  const validType = formatTypeMap.get(format);
+
+  if (validType === undefined) {
+    // Ignore assertions for module types we don't recognize, to allow new
+    // formats in the future.
+    return true;
+  } else if (validType === kImplicitAssertType) {
+    if (!importAssertions ||
+        !ObjectPrototypeHasOwnProperty(importAssertions, 'type')) {
+      return true;
+    }
+    throw new ERR_IMPORT_ASSERTION_TYPE_UNSUPPORTED(format,
+                                                    importAssertions.type);
+  } else if (validType === importAssertions?.type) {
+    return true;
+  } else if (!ObjectPrototypeHasOwnProperty(importAssertions, 'type')) {
+    throw new ERR_IMPORT_ASSERTION_TYPE_FAILED(format, validType);
+  } else {
+    throw new ERR_IMPORT_ASSERTION_TYPE_FAILED(format, validType);
+  }
```

Comment:
So the “before” version of this has an advantage over the proposed alternative in that it covers the success cases first: notice how the `return true` lines are as close to the top as they can be. Most code _will_ validate successfully, as generally the only time these errors will likely be thrown is during development when the author forgets to add an assertion; so we want to optimize for the most likely scenarios and return early for those happy paths.

Could we do without a few `else`s? Sure. I’m not sure that’s necessarily an improvement, though, as the overall `if`/`else` structure makes it clear that this is all one logic block rather than separate ones. I can be persuaded if others feel otherwise, however, that part feels more stylistic to me.

Reply:
I don't think that's stylistic. it just seems unnecessary to me. there's a lot of boilerplate one can write where the outcome will be the same at runtime. I thought this would be not even controversial. there's even an eslint rule for this scenario as well as endless talks about unnecessary and redundant `else if` blocks.

> So the “before” version of this has an advantage over the proposed alternative in that it covers the success cases first: notice how the return true lines are as close to the top as they can be. 

This is anything but a hot path. it makes more sense from a logical point of view to check something for existence and use it in a later condition to compare it to something. this is also my typescript mind speaking.

anyhow, I thought it would be a good suggestion. I personally like concise meaningful code, specially for maintenance reasons and anyone reading it at a later point in time.

Reply:
> anyhow, I thought it would be a good suggestion. I personally like concise meaningful code, specially for maintenance reasons and anyone reading it at a later point in time.

I don't think it's a bad suggestion, but I also think the current version is actually more concise and meaningful 🤔 Your suggestion separates each condition as separate blocks, while I for one thinks it's easier to think of it a one block (your suggestion ends up being more lines than the current version), but I think the stylistic part of it is really subjective.

Regading the re-ordering to make `ObjectPrototypeHasOwnProperty(importAssertions, 'type')` check before checking it's value, I agree that's definitely something we should do.

Reply:
Please see https://github.com/nodejs/node/commit/ddfeb1d1770089280d53b226256133dffab8cc91.